---
title: "Staphopia tutorial"
output:
  html_document:
    toc: true
    theme: united
---

# staphopia-r example commands

Load libraries used in this tutorial. (you may have to download and install)

```{r}
library(staphopia)
library(dplyr)
library(Biostrings)
library(phangorn)
library(purrr)
library(ape)
```

# Creating a sample set

There are multiple ways to gather a set of genomes ("samples" in Staphopia) for analysis.  There are some pre-set collections of samples gathered under "tags".  A list of tags can be generated by ```get_all_tags```

```{r}
tags <- get_all_tags()
```

The difference between ```get_all_tags`` and ```get_public_tags``` is that the former command could include private samples that have not yet been submitted to NCBI.  We use "public" to mean that that the data has been submitted to NCBI/ENA and is freely available for download.


```{r}
redrun <- get_samples_by_tag(10)
head()
```

An alternative method is to get all the public samples in the database.  This requires a staphopia account/token.  Obviously, this disadvantage of this large set is that it can slow down analysis.

```{r}
pub_sam <- get_public_samples()
```

list the ST8 projects
```{r}
ST8_samples <- filter(pub_sam,st_stripped == 8) %>%
  select(sample_id,sample_tag,st_stripped)
```

# Exploring SNPs and Indels

Based on N315 reference genome coordinates
```{r}
graR_SNPs <- get_snps_in_bulk((708245*3):(708919*3))
rpoB_SNPs <- get_snps_in_bulk((579620*3):(583171*3))
walK_SNPs <- get_snps_in_bulk((25648*3):(27474*3))
```

look at one specific SNP
```{r}
snp2124752 <- get_samples_by_snp(graR_SNPs$id[33])
pub_sam %>% filter(sample_id %in% snp2124752)
```

count SNPs
```{r}
graR_SNP_Counts <- map_int(graR_SNPs$id,function(x) length(get_samples_by_snp(x)))   
```

add counts of SNPs to columns
```{r}
graR_ <- cbind(graR_SNPs,graR_SNP_Counts) 
```

barplot
```{r}
barplot(graR_SNP_Counts, main = "graR SNP distribution")
```
# Genes

get gene products
```{r}
gp <- get_gene_products()
```

IS256 transposase is id 2034
```{r}
#
IS256 <- get_genes(ST8_samples$sample_id,product_id=2034)

```

create stringset
```{r}
IS256_out <- DNAStringSet(IS256$dna)
names(IS256_out) <- paste(IS256$sample_id,IS256$id,sep = "_")
IS256_out <- IS256_out[width(IS256_out)==1173]
```

phylogeny/ heat map

```{r}
IS256_bin <- as.DNAbin(IS256_out)
dm <- dist.dna(IS256_bin)
no_SNps <- as.matrix(as.integer(dm*1173))
hist(no_SNps)
njtree <- nj(dm)
temp <- as.matrix(dm)
heatmap(temp,cexRow = 0.2, cexCol = 0.2)
njtree <- midpoint(njtree)
plot(njtree,show.tip.label = T, cex = 0.5)
```

## Writing data to hard drive


# Metadata
```{r}
ST8_meta <- get_metadata(ST8_samples$sample_id)
```

# Genome assembly and sequence quality

# SCCmec, MRSA and antibiotic resistance

# Virulence genes

# Rapid core genome trees

# MLST and cgMLST


